File: go/groupAnagrams.go
// https://leetcode.com/problems/group-anagrams

package leetcode

func groupAnagrams(strs []string) [][]string {
	anagrams := make(map[[26]int][]string)

	for _, str := range strs {
		var key [26]int
		for _, c := range str {
			key[c-'a']++
		}
		anagrams[key] = append(anagrams[key], str)
	}

	res := make([][]string, 0, len(anagrams))
	for _, val := range anagrams {
		res = append(res, val)
	}

	return res
}

File: python/src/groupAnagrams.py
# https://leetcode.com/problems/group-anagrams/

from collections import defaultdict
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams = defaultdict(list)
        for word in strs:
            key = [0] * 26
            for c in word:
                key[ord(c) - ord("a")] += 1
            anagrams[tuple(key)].append(word)
        return anagrams.values()

    def groupAnagrams_sort(self, strs: List[str]) -> List[List[str]]:
        anagrams = defaultdict(list)
        for word in strs:
            sorted_word = "".join(sorted(word))
            anagrams[sorted_word].append(word)
        return anagrams.values()
